### ISSUE TITLE
Memory exhausted assigning concatenated strings to list item

### ISSUE THREAD

**Issue by tomas-mbit** [Microsoft/pxt-microbit#1162]
**Describe the bug**

Concatenating two strings and assigning them to a position in a list exhausts memory after a few hundred iterations.

**To Reproduce**

This program

![bild](https://user-images.githubusercontent.com/42862513/44949911-6e33eb80-ae3b-11e8-8574-4572998dc8b7.png)

https://makecode.microbit.org/_MJKFi8DcRdfa

After some 330 iterations, a Microbit 020 error occurs. Making the strings to concatenate longer makes memory run out faster and vice versa. Relationship is not quite inversely proportional. Seems like the contents of the strings plus a little more is left in memory for each iteration.  Variants of the program sometimes cause an 030 error or cause the microbit to hang.

If a single string is assigned rather than the concatenation of two strings, there is no problem. 
If the concatenation of two strings is assigned to a string variable instead of to a list, there is also no problem.

**Expected behavior**

Memory should not run out.

**Desktop (please complete the following information):**
 - OS: Windows 7
 - Browser Firefox
 - Version 61.0.2 (64-bitars)

--Tomas

**Comment by pelikhan** [comment]
Correct we should not OOM. This needs to be investigated.
```
let Count = 0
let Testlist: string[] = []
Testlist = [""]
Count = 0
serial.redirectToUSB()
basic.forever(() => {
    for (let i = 0; i < 10; i++) {
        Testlist[0] = "13 characters" + "13 characters"
        Count += 1
    }
    serial.writeNumber(Count)
    serial.writeLine("")
})
```

**Comment by tomas-mbit** [comment]
Screenshot in English 
/Tomas

![bild](https://user-images.githubusercontent.com/42862513/45543576-263a9e80-b815-11e8-9a5b-733c9f0913f9.png)

**Comment by pelikhan** [comment]
Sounds like a ref counting issue somewhere.

**Comment by mmoskal** [comment]
Does it repro on v1?

**Comment by pelikhan** [comment]
@tomas-mbit this issue seems to be fixed in https://makecode.microbit.org/v1#pub:_MJKFi8DcRdfa .

**Comment by tomas-mbit** [comment]
The example seems to work.

An relevant question is of course if the new code in v1 fixes the error, or if the error is still there but because of changes to the code, it is no longer exposed in this particular setting. 

I ran the linked program. 500 000 iterations and no problem. (I think I might have changed the printout interval to 100). However, when I tried my original large program where I first ran into this problem, it seemed to work fine at first, but after running overnight it had slowed down and was hardly moving at all. 
I made a quick attempt to narrow it down to something short, but didn't succeed. I might have some  time for more tries during the coming week.
I can supply the larger program if you want, but it's somewhat messy and uncommented. Size is about 250 Javascript lines.

I had some problems loading the old hex file into v1. It worked at first, but then caused the web browser to hang (used Chrome this time). 

--Tomas

(After first reporting this bug, I made a version of the large program where almost all list/array handling was removed. This version was very stable and has run overnight without any problems. )

**Comment by mmoskal** [comment]
It's quite easy to run out of memory on the microbit. The slowdown is probably something else though. Please do post the program, maybe something clear will jump out.

**Comment by tomas-mbit** [comment]
Here's the program.  https://makecode.microbit.org/_MhE71zgfUEFT
(It's too large for a meaningful sceenshot.)

I tried it again today after a slight modification to remove an alternative possible source of slowdown;  after 10 hours it was almost not moving at all.
It's a networking program so you have to run it on two microbits simultaneously. By pressing A/B at the start sequence you select (among other things, please look at the startup code) one microbit to be node 1 and the other to be node 2. Then they send the same data over and over to each other and print out what they receive. There's acknowledgement/retransmission and packet fragmentation/reassembly. 
There's a layer 1 header consisting of 1 digit sender id and 3 digits layer 1 serial number. This is intended for relay microbits (not tested yet) that will simply retransmit any message whose sender/serial they haven't already seen.
Layer 2 has a 1 digit destination id, a 1 digit sender id, and either a 3 digit serial number and a payload or "ACK" and the serial number of the acknowledged packet.  For simplicity, only one layer 2 message is outstanding; only when a message is ACK:ed will the next be sent. 

It's a bit tricky to grasp what's going on, there seems to be a bit of packet collisions when the two microbits come into "destructive sync", transmitting at the same time despite the small random delay in the transmit code. Inserting printouts into the code changes the timing. Nevertheless, the slowdown over 10 hours is so drastic that it smells of something possibly wrong somewhere. 

A possible thing to try could be to remove the radio part and insert code that simply mimics the other microbit. But this too changes the timing of course.

--Tomas

**Comment by tomas-mbit** [comment]
I've fiddled some more with this, but I am getting inconsistent results. 
I tried selectively removing parts of the program. Without assigning joined strings to list items everything was ok.  With lists, but without radio, it was also ok. Then, going back to the original, it was slow even from the beginning. The microbits were positioned on each side of my laptop, but when I put them within a few centimeters from one another, speed went back to normal. After explicitly setting transmit power  things returned to normal. Both maximum power (7) and minimum power (0) worked ok.  Maybe the slowdown is a different problem.
What transmit power is used if it is not set explicitly? Is there some kind of adaptive setting?
--Tomas
